libbeat/filter/condition.go

```diff
@@ -0,0 +1,325 @@
+package filter
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+type Condition struct {
+	Equals   map[string]EqualsValue
+	Contains map[string]string
+	Regexp   map[string]*regexp.Regexp
+	Range    map[string]RangeValue
+}
+
+func NewCondition(config ConditionConfig) (*Condition, error) {
+
+	c := Condition{}
+	c.Equals = map[string]EqualsValue{}
+	c.Contains = map[string]string{}
+	c.Regexp = map[string]*regexp.Regexp{}
+	c.Range = map[string]RangeValue{}
+
+	if err := c.AddEquals(config.Equals); err != nil {
+		return nil, err
+	}
+	if err := c.AddContains(config.Contains); err != nil {
+		return nil, err
+	}
+	if err := c.AddRegexp(config.Regexp); err != nil {
+		return nil, err
+	}
+	if err := c.AddRange(config.Range); err != nil {
+		return nil, err
+	}
+
+	return &c, nil
+}
+
+func (c *Condition) AddEquals(equals map[string]string) error {
+
+	for field, value := range equals {
+
+		i, err := strconv.Atoi(value)
+		if err == nil {
+			c.Equals[field] = EqualsValue{Int: i}
+		} else {
+			c.Equals[field] = EqualsValue{Str: value}
+		}
+	}
+	return nil
+}
+
+func (c *Condition) AddContains(contains map[string]string) error {
+
+	c.Contains = contains
+	return nil
+}
+
+func (c *Condition) AddRegexp(r map[string]string) error {
+
+	for field, value := range r {
+		reg, err := regexp.CompilePOSIX(value)
+		if err != nil {
+			return err
+		}
+		c.Regexp[field] = reg
+	}
+	return nil
+}
+
+func (c *Condition) AddRange(config map[string]RangeValue) error {
+
+	for field, rangeValue := range config {
+		c.Range[field] = rangeValue
+	}
+	return nil
+}
+
+func (c *Condition) Check(event common.MapStr) bool {
+
+	if !c.CheckEquals(event) {
+		return false
+	}
+	if !c.CheckContains(event) {
+		return false
+	}
+	if !c.CheckRegexp(event) {
+		return false
+	}
+	if !c.CheckRange(event) {
+		return false
+	}
+
+	return true
+}
+
+func (c *Condition) CheckEquals(event common.MapStr) bool {
+
+	for field, equalValue := range c.Equals {
+
+		value, err := event.GetValue(field)
+		if err != nil {
+			logp.Debug("filter", "unavailable field %s: %v", field, err)
+			return false
+		}
+
+		switch value.(type) {
+		case uint8, uint16, uint32, uint64, int8, int16, int32, int64, int, uint:
+			return value == equalValue.Int
+		case string:
+			return value == equalValue.Str
+		default:
+			logp.Warn("unexpected type %T in equals condition as it accepts only integers and strings. ", value)
```

Comment:
Perhaps it makes even more sense to return here an error instead of only logging it. Makes it clearer what happened and easier to test. See https://github.com/elastic/beats/pull/1395/files#diff-d5fd542a1e55817820849e3fd1f03e30R72 check4.